core/src/main/antlr4/com/databricks/labs/remorph/parsers/snowflake/SnowflakeParser.g4

```diff
@@ -3321,17 +3334,17 @@ withExpression: WITH commonTableExpression (COMMA commonTableExpression)*
     ;
 
 commonTableExpression
-    : tableName = id (L_PAREN columns += id (COMMA columns += id)* R_PAREN)? AS L_PAREN selectStatement setOperators* R_PAREN
+    : tableName = id (L_PAREN columns += id (COMMA columns += id)* R_PAREN)? AS L_PAREN ((selectStatement setOperators*) | expr) R_PAREN
     ;
 
 selectStatement
     : selectClause selectOptionalClauses limitClause?
     | selectTopClause selectOptionalClauses //TOP and LIMIT are not allowed together
+    | L_PAREN selectStatement R_PAREN
     ;
```

Comment:
What is the reason to add the `L_PAREN` `R_PAREN` here? It means that it can be `(((( SELECT ... ))))` etc.

It is probably too much work now, but the parser should not concern itself with disallowing TOP and LIMIT together as that will cause a somewhat meaningless syntax error, where we would want (in a parser rather than a a translator perhaps), something like `"TOP and LIMIT are mutually exclusive in a SELECT statement"`.